qualtran/bloqs/block_encoding.py

```diff
@@ -182,13 +194,66 @@ def pretty_name(self) -> str:
         return 'Prep'
 
 
+class BlockEncoding(Bloq):
+    r"""Abstract interface for an arbitrary block encoding.
+
+    A $(\alpha, a, \epsilon) block encoding of an s-qubit operator $H$ if it obeys:
+
+    $$
+        \equiv \lVert H - \alpha (\langle G|_a\otimes I_s U |G\rangle_a \otimes I_s) \rVert
+        \le \epsilon,
+    $$
+
+    where $a$ is an ancilla register and $s$ is the system register, $U$ is a unitary sometimes
+    called a signal oracle and encodes $H$ in its top right corner, while $|G\rangle_a$ is
+    sometimes called the signal state.
+
+    Developers users must implement a method to return a bloq preparing the state $|G\rangle$.
+
+    Users must specify:
+        1. the normalization constant $\alpha \ge \lVert A \rVert$, where
+            $\lVert \cdot \rVert denotes the spectral norm.
+        2. the precision to which the block encoding is to be prepared ($\epsilon$).
+
+    References:
+        [Hamiltonian Simulation by Qubitization](https://quantum-journal.org/papers/q-2019-07-12-163/)
+            Sec 2 and 3 for introduction and definition of terms.
+
+        [The power of block-encoded matrix powers: improved regression techniques via faster Hamiltonian simulation](https://arxiv.org/abs/1804.01973)
+            Definition 3 page 8.
+
+    """
+    alpha: SymbolicFloat
+    epsilon: SymbolicFloat
+
+    @property
+    @abc.abstractmethod
+    def selection_bitsize(self) -> int:
+        """The bitsize for the register `a` registers above"""
+
+    @property
+    @abc.abstractmethod
+    def junk_bitsize(self) -> int:
+        """An additional junk register typically used for Prepare which is not reflected on."""
```

Comment:
Yeah it helps you combine many different registers (the selection registers for Prepare might have 10 different wires) into a single register of the same total bitsize. It's helpful for writing higher level algorithms which don't care about the details of these 10 registers. Like the chebyshev polynomial or the block encoding which would just be defined on ancilla (+junk) and system registers, decomposing further (zooming in) would reveal the details of the big registers.

I can do that sure.

Reply:
I think we want to define the block encoding to be high level so that the signature should be 3 registers for system, junk and selection rather than a list of registers and use the partition block as is.